frontends/p4/def_use.h

```diff
@@ -555,63 +656,9 @@ class ComputeWriteSet : public Inspector, public IHasDbPrint {
-          returnedDefinitions(nullptr),
-          exitDefinitions(new Definitions()),
-          storageMap(allDefinitions->storageMap),
-          lhs(false),
-          virtualMethod(false) {
-        CHECK_NULL(allDefinitions);
-        visitDagOnce = false;
-    }
-
-    // expressions
-    bool preorder(const IR::Literal *expression) override;
-    bool preorder(const IR::Slice *expression) override;
-    bool preorder(const IR::TypeNameExpression *expression) override;
-    bool preorder(const IR::PathExpression *expression) override;
-    bool preorder(const IR::Member *expression) override;
-    bool preorder(const IR::ArrayIndex *expression) override;
-    bool preorder(const IR::Operation_Binary *expression) override;
-    bool preorder(const IR::Mux *expression) override;
-    bool preorder(const IR::SelectExpression *expression) override;
-    bool preorder(const IR::ListExpression *expression) override;
-    bool preorder(const IR::Operation_Unary *expression) override;
-    bool preorder(const IR::MethodCallExpression *expression) override;
-    bool preorder(const IR::DefaultExpression *expression) override;
-    bool preorder(const IR::Expression *expression) override;
-    bool preorder(const IR::InvalidHeader *expression) override;
-    bool preorder(const IR::InvalidHeaderUnion *expression) override;
-    bool preorder(const IR::P4ListExpression *expression) override;
-    bool preorder(const IR::HeaderStackExpression *expression) override;
-    bool preorder(const IR::StructExpression *expression) override;
-    // statements
-    bool preorder(const IR::P4Parser *parser) override;
-    bool preorder(const IR::P4Control *control) override;
-    bool preorder(const IR::P4Action *action) override;
-    bool preorder(const IR::P4Table *table) override;
-    bool preorder(const IR::Function *function) override;
-    bool preorder(const IR::AssignmentStatement *statement) override;
-    bool preorder(const IR::ReturnStatement *statement) override;
-    bool preorder(const IR::ExitStatement *statement) override;
-    bool preorder(const IR::BreakStatement *statement) override;
-    bool handleJump(const char *tok, Definitions *&defs);
-    bool preorder(const IR::ContinueStatement *statement) override;
-    bool preorder(const IR::IfStatement *statement) override;
-    bool preorder(const IR::ForStatement *statement) override;
-    bool preorder(const IR::ForInStatement *statement) override;
-    bool preorder(const IR::BlockStatement *statement) override;
-    bool preorder(const IR::SwitchStatement *statement) override;
-    bool preorder(const IR::EmptyStatement *statement) override;
-    bool preorder(const IR::MethodCallStatement *statement) override;
-
-    const LocationSet *writtenLocations(const IR::Expression *expression) {
-        expression->apply(*this);
-        return getWrites(expression);
-    }
+ private:
+    // TODO: Make absl::flat_hash_set instead?
+    std::unordered_set<loc_t> &cached_locs;
```

Comment:
or use an `hvec_set`

Reply:
I don't think there is any `hvec_set` in the tree

Reply:
You need to ensure stable addresses of values around insertions. So, `flat_hash_set` is not an option here. You might want to give `node_hash_set` a try:
```suggestion
    absl::node_hash_set<loc_t, Util::Hash> &cached_locs;
```

Reply:
We are not doing any iteration of `cached_locs`, so insertion order does not matter here. If you think `absl::node_hash_set` would be better, feel free to push your changes to this branch as I ran into some errors when trying your suggestion and I'd rather not spend the time debugging this unless there is a good reason to do so.

Reply:
It is not about stable iteration order. The issue is much more subtle, but is *very* important: you are doing (though I would prefer `->first` here):

```c++
    return &*cached_locs.insert(tmp).first;
```
Note that you are returning the address of hash map entry. Therefore you need to ensure this address will not change during insertions / deletions. STL unordered containers guarantee this (as each bucket is essentially a list). `flat_hash_set` does not guarantee this, `node_hash_set` does. See https://abseil.io/docs/cpp/guides/container for more information

Reply:
> Note that you are returning the address of hash map entry. Therefore you need to ensure this address will not change during insertions / deletions.

Ok, I see what you mean now.

I'm not convinced that using `std::unordered_set` is a problem in the first place, so I will not spend the time trying to get `node_hash_set` to work, but feel free to push such changes to this branch if you know how to do it and think that it will improve the performance.

(but I will at least update the comment if you don't want to make these changes)